lib/signs_ui/config/sign_group_to_sign_configs.ex

```diff
@@ -0,0 +1,104 @@
+  Signs unchanged. When Sign Groups are changed, those changes are propagated to the underlying Sign
+  configs. There are three kinds of changes:
+
+  * New Sign Group - all the signs in this group have their custom text, mode and expiration options
+    set to the same values as the group value.
+  * Sign Group Deleted - all the signs in the group are reset back to Auto
+  * Sign Group Updated - Same as New Sign Group, plus any signs that were removed are reset back to
+    Auto.
+  """
+
+  alias SignsUi.Config.Sign
+  alias SignsUi.Config.SignGroup
+  alias SignsUi.Config.SignGroups
+  alias SignsUi.Config.State
+
+  @doc """
+  Transforms the Sign Group changes into a map of Sign Config updates.
+  """
+  @spec apply(SignGroups.t(), State.t()) :: %{Sign.id() => Sign.t()}
+  def apply(sign_group_changes, config_state) do
+    Enum.reduce(sign_group_changes, %{}, fn {sign_group_id, sign_group}, acc ->
+      Map.merge(
+        process_sign_group(config_state.sign_groups[sign_group_id], normalize(sign_group)),
+        acc
+      )
+    end)
+  end
+
+  @spec process_sign_group(SignGroup.t() | nil, SignGroup.t() | nil) :: %{Sign.id() => Sign.t()}
+  defp process_sign_group(old_sign_group, new_sign_group)
+
+  # Group is being deleted
+  defp process_sign_group(%SignGroup{} = old_group, nil) do
+    set_signs_to_auto(old_group.sign_ids)
+  end
+
+  # Group is being added
+  defp process_sign_group(nil, %SignGroup{} = new_sign_group) do
+    set_signs_to_static_text(
+      new_sign_group.sign_ids,
+      new_sign_group.line1,
+      new_sign_group.line2,
+      new_sign_group.expires,
+      new_sign_group.alert_id
+    )
+  end
+
+  # Group is being updated
+  defp process_sign_group(%SignGroup{} = old_group, %SignGroup{} = new_group) do
+    additions =
+      set_signs_to_static_text(
+        new_group.sign_ids,
+        new_group.line1,
+        new_group.line2,
+        new_group.expires,
+        new_group.alert_id
+      )
+
+    removed_signs =
+      MapSet.difference(
+        MapSet.new(old_group.sign_ids),
+        MapSet.new(new_group.sign_ids)
+      )
+
+    removals = set_signs_to_auto(removed_signs)
+
+    Map.merge(additions, removals)
+  end
+
+  # Shouldn't occur, but for completeness
+  defp process_sign_group(nil, nil) do
+    %{}
+  end
+
+  defp set_signs_to_auto(sign_ids) do
+    Map.new(sign_ids, fn sign_id ->
+      {sign_id, %Sign{id: sign_id, config: %{mode: :auto}}}
+    end)
+  end
+
+  defp set_signs_to_static_text(sign_ids, line1, line2, expires, alert_id) do
+    Map.new(sign_ids, fn sign_id ->
+      {
+        sign_id,
+        %Sign{
+          id: sign_id,
+          config: %{
+            mode: :static_text,
+            line1: line1,
+            line2: line2,
+            expires: expires,
+            alert_id: alert_id
+          }
+        }
+      }
+    end)
+  end
+
+  defp normalize(%SignGroup{} = sign_group), do: sign_group
+  defp normalize(sg) when sg == %{}, do: nil
```

Comment:
Any reason not to just put the `when sign_group == %{}` guard in the `process_sign_group` clauses themselves? Then you could ditch this and maybe also the "shouldn't occur" clause, and simplify the typespec.

Reply:
I had it that way initially, but I felt that `%SignGroup{} | nil` (i.e.: `Option<SignGroup>`) was a better way of representing what the data was. The empty struct isn't semantically an empty struct there so much as a kind of flag.

We'd still need the "shouldn't occur" clause since `%{}` is not a subtype of `%SignGroup{}` and the pattern match fails.